driver-core/src/main/com/mongodb/operation/ChangeStreamBatchCursor.java

```diff
@@ -0,0 +1,164 @@
+/*
+ * Copyright 2017 MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.mongodb.operation;
+
+import com.mongodb.Function;
+import com.mongodb.MongoNamespace;
+import com.mongodb.MongoNotPrimaryException;
+import com.mongodb.MongoQueryException;
+import com.mongodb.MongoServerException;
+import com.mongodb.ServerAddress;
+import com.mongodb.ServerCursor;
+import com.mongodb.binding.ReadBinding;
+import org.bson.BsonArray;
+import org.bson.BsonDocument;
+import org.bson.BsonInt64;
+import org.bson.BsonString;
+import org.bson.RawBsonDocument;
+import org.bson.codecs.BsonDocumentCodec;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import static com.mongodb.operation.CommandOperationHelper.executeWrappedCommandProtocol;
+import static java.util.Collections.singletonList;
+
+final class ChangeStreamBatchCursor<T> implements BatchCursor<T> {
+    private static final BsonDocumentCodec BSON_DOCUMENT_CODEC = new BsonDocumentCodec();
+    private final ReadBinding binding;
+    private final ChangeStreamOperation<T> changeStreamOperation;
+
+    private volatile BsonDocument resumeToken;
+    private volatile QueryBatchCursor<RawBsonDocument> wrapped;
+
+    ChangeStreamBatchCursor(final ChangeStreamOperation<T> changeStreamOperation, final ReadBinding binding,
+                            final QueryBatchCursor<RawBsonDocument> wrapped) {
+        this.changeStreamOperation = changeStreamOperation;
+        this.resumeToken = changeStreamOperation.getResumeToken();
+        this.binding = binding;
```

Comment:
Since the binding is stored in a property, it should be retained here and released in the close method.  There's no guarantee that anyone else is holding on to a reference to it, so it could go to zero.  In this case the underlying QueryBatchCursor retains a reference to the ConnectionSource, which retains one to its binding, but that's not guaranteed.